pkg/controller/kubelet-config/kubelet_config_features.go

```diff
@@ -231,3 +209,87 @@ func generateFeatureMap(features *osev1.FeatureGate, exclusions ...string) (*map
 	}
 	return &rv, nil
 }
+
+func generateKubeConfigIgnFromFeatures(cc *mcfgv1.ControllerConfig, templatesDir, role string, featureGates map[string]bool) ([]byte, error) {
+	// Generate the original KubeletConfig
+	originalKubeletIgn, err := generateOriginalKubeletConfig(cc, templatesDir, role, nil)
+	if err != nil {
+		return nil, err
+	}
+	if originalKubeletIgn.Contents.Source == nil {
+		return nil, fmt.Errorf("could not find original Kubelet config to decode")
+	}
+	dataURL, err := dataurl.DecodeString(*originalKubeletIgn.Contents.Source)
+	if err != nil {
+		return nil, err
+	}
+	originalKubeConfig, err := decodeKubeletConfig(dataURL.Data)
+	if err != nil {
+		return nil, err
+	}
+	// Check to see if FeatureGates are equal
+	if reflect.DeepEqual(originalKubeConfig.FeatureGates, featureGates) {
+		// When there is no difference, this isn't an error, but no machine config should be created
+		return nil, nil
```

Comment:
I'm not sure if this will work. Upon an upgrade, all MCs should be updated with at least the new controller version. If the featuregate doesn't change, we still have to return the same MC such that the sync funtion above doesn't hit the

```
if rawCfgIgn == nil {
	continue
}
```
which will not update the
```
mc.ObjectMeta.Annotations = map[string]string{
	ctrlcommon.GeneratedByControllerVersionAnnotationKey: version.Hash,
}
```
which then in turn causes the MCO to fail because of a rendered version mismatch for the controller "(i.e. MCC sees that a controller-created config never got updated). So we should still return the same config so the above sync happens.

Have you tried installing (from existing releases), adding a feature gate and then upgrading to this PR? That should be a good check.

Reply:
The logic here hasn't changed as far as I understand it, the idea of this function was just to extract the logic that was already in place, but not change the logic.

Before, there was this same check that caused a `continue` in the loop, I've just replaced this with an early return and the continue stays in the top of the loop.

I think the reason this works is because we are comparing the default feature gates (ie if you didn't have a `FeatureGate`) with what is in the `FeatureGate` in the cluster, if it exists. Once you've added a feature gate, you can't remove it or modify it to remove any feature that has been enabled. So if a MC has been generated because of the FG, it will always be the case that these differ and it will always cause the MC to be updated.

Also note that if you have a `FeatureGate` in the cluster, you can't upgrade your cluster, so the version should never change anyway if this logic has been triggered.

Reply:
Oh I see, defaultFeatures is static, ok that makes sense I think.

When we say "can't upgrade" we mean we cannot upgrade y-stream correct? Based on the CI job it seems featuregates are blocked from upgrading at the y-stream level but not at the z-stream level, since it makes use of upgradeable=false. Just wanted to check that understanding